ca/ca_test.go

```diff
@@ -207,261 +194,527 @@ func setup(t *testing.T) *testCtx {
-		err = cert.CheckSignatureFrom(issuer.Cert.Certificate)
-		if err == nil {
-			validated = true
-			break
-		}
+
+	ca, err := newCAArgs(t).make()
+	if err != nil {
+		t.Fatalf("making test ca: %s", err)
 	}
-	test.Assert(t, validated, "Certificate failed signature validation")
-	test.AssertMetricWithLabelsEquals(t, ca.metrics.signatureCount, prometheus.Labels{"purpose": "precertificate", "status": "success"}, 1)
 
-	// Test that an ECDSA CSR gets issuance from an ECDSA issuer.
-	issuedCertDER, err = ca.issuePrecertificate(ctx, profile, &capb.IssueCertificateRequest{Csr: ECDSACSR, RegistrationID: mrand.Int63(), OrderID: mrand.Int63(), CertProfileName: "legacy"})
-	test.AssertNotError(t, err, "Failed to issue certificate")
-	cert, err = x509.ParseCertificate(issuedCertDER)
-	test.AssertNotError(t, err, "Certificate failed to parse")
-	validated = false
-	for _, issuer := range ca.issuers.byAlg[x509.ECDSA] {
-		err = cert.CheckSignatureFrom(issuer.Cert.Certificate)
-		if err == nil {
-			validated = true
-			break
-		}
+	for _, tc := range []struct {
+		profile      string
+		wantValidity time.Duration
+	}{
+		{
+			profile:      "legacy",
+			wantValidity: 90 * 24 * time.Hour,
+		},
+		{
+			profile:      "modern",
+			wantValidity: 6 * 24 * time.Hour,
+		},
+	} {
+		t.Run(tc.profile, func(t *testing.T) {
+			t.Parallel()
+
+			res, err := ca.IssueCertificate(t.Context(), &capb.IssueCertificateRequest{
+				RegistrationID: 1, OrderID: 1,
+				Csr: ECDSACSR, CertProfileName: tc.profile,
+			})
+			if err != nil {
+				t.Fatalf("IssueCertificate(profile=%q) = %q, but want success", tc.profile, err)
+			}
+
+			cert, err := x509.ParseCertificate(res.DER)
+			if err != nil {
+				t.Fatalf("parsing certificate: %s", err)
+			}
+
+			// We use the validity period as a proxy for detecting whether the correct
+			// profile was selected and used, since we know that the validity period
+			// differs between the two test profiles.
+			validity := cert.NotAfter.Add(time.Second).Sub(cert.NotBefore)
+			if validity != tc.wantValidity {
+				t.Errorf("IssueCertificate(profile=%q) = validity %d, but want %d", tc.profile, validity, tc.wantValidity)
+			}
+		})
 	}
-	test.Assert(t, validated, "Certificate failed signature validation")
-	test.AssertMetricWithLabelsEquals(t, ca.metrics.signatureCount, prometheus.Labels{"purpose": "precertificate", "status": "success"}, 2)
 }
 
-func TestUnpredictableIssuance(t *testing.T) {
-	testCtx := setup(t)
-	sa := &mockSA{}
+func TestIssueCertificate_IssuerSelection(t *testing.T) {
+	t.Parallel()
+	cargs := newCAArgs(t)
+	origIssuers := cargs.boulderIssuers
+
+	ca, err := cargs.make()
+	if err != nil {
+		t.Fatalf("making test ca: %s", err)
+	}
+
+	for _, tc := range []struct {
+		name        string
+		csr         []byte
+		wantIssuers []*issuance.Issuer
+		wantKUs     x509.KeyUsage
+	}{
+		{
+			name:        "ECDSA",
+			csr:         ECDSACSR,
+			wantIssuers: origIssuers[2:],
```

Comment:
This creates some spooky action at a distance, depending on specific properties of the list of issuers created by newCAArgs. If we change that list, it might be surprising to get errors. Fancy fix would be to filter that list by key type; but probably a simple fix is better here: adding a comment above each `wantIssuers` assignment saying e.g. "Assumption: in newCAArgs, the only ECDSA issuers are the last two" / "Assumption: in newCAArgs, the only RSA issuers are the first two"